airflow/jobs/scheduler_job.py

```diff
@@ -823,14 +823,30 @@ def _create_dag_runs(self, dag_models: Iterable[DagModel], session: Session) ->
         existing_dagruns = (
             session.query(DagRun.dag_id, DagRun.execution_date).filter(existing_dagruns_filter).all()
         )
+        max_queued_dagruns = conf.getint('scheduler', 'max_queued_runs_per_dag')
+
+        queued_runs_of_dags = defaultdict(
+            int,
+            session.query(DagRun.dag_id, func.count('*'))
+            .filter(  # We use `list` here because SQLA doesn't accept a set
+                # We use set to avoid duplicate dag_ids
+                DagRun.dag_id.in_(list({dm.dag_id for dm in dag_models})),
+                DagRun.state == State.QUEUED,
+            )
+            .group_by(DagRun.dag_id)
+            .all(),
+        )
 
         for dag_model in dag_models:
+            # Lets quickly check if we have exceeded the number of queued dagruns per dags
+            total_queued = queued_runs_of_dags[dag_model.dag_id]
+            if total_queued >= max_queued_dagruns:
+                continue
```

Comment:
```suggestion
            if total_queued >= max_queued_dagruns:
				self.log.debug("DAG %s is at max_queued_runs_per_dag (%d), not creating any more queued runs", dag_model.dag_id, max_queued_dagruns)
                continue
```

I'm not sure if this will get too noisy, but it might be worthwhile to have some logging here.

Reply:
It will get noisy. Imagine possible dag_run to create is up to 2000, this would print ~1984 times

Reply:
Debug is off by default, so we could have this on only at debug level without overwhelming people in normal cases.